tuned/exports/unix_socket_exporter.py

```diff
@@ -0,0 +1,195 @@
+	bound to the original object instance. While the wrapper will be bound
+	to an object we dynamically construct.
+	"""
+
+	def __init__(self, socket_path=consts.CFG_DEF_UNIX_SOCKET_PATH,
+				 signal_paths=consts.CFG_DEF_UNIX_SOCKET_SIGNAL_PATHS,
+				 ownership=consts.CFG_DEF_UNIX_SOCKET_OWNERSHIP,
+				 permissions=consts.CFG_DEF_UNIX_SOCKET_PERMISIONS):
+
+		self._socket_path = socket_path
+		self._socket_object = None
+		self._socket_signal_paths = re.split(r",;", signal_paths) if signal_paths else []
+		self._socket_signal_objects = []
+		self._ownership = [-1, -1]
+		if ownership:
+			ownership = ownership.split()
+			for i, o in enumerate(ownership[:2]):
+				try:
+					self._ownership[i] = int(o)
+				except ValueError:
+					try:
+						# user
+						if i == 0:
+							self._ownership[i] = pwd.getpwnam(o).pw_uid
+						# group
+						else:
+							self._ownership[i] = grp.getgrnam(o).gr_gid
+					except KeyError:
+						log.error("%s '%s' does not exists, leaving default" % ("User" if i == 0 else "Group", o))
+		self._permissions = permissions
+
+		self._unix_socket_methods = {}
+		self._signals = set()
+		self._thread = None
+		self._conn = None
+		self._channel = None
+
+	def running(self):
+		return self._thread is not None
+
+	def export(self, method, in_signature, out_signature):
+		if not ismethod(method):
+			raise Exception("Only bound methods can be exported.")
+
+		method_name = method.__name__
+		if method_name in self._unix_socket_methods:
+			raise Exception("Method with this name (%s) is already exported." % method_name)
+
+		class wrapper(object):
+			def __init__(self, in_signature, out_signature):
+				self._in_signature = in_signature
+				self._out_signature = out_signature
+				
+			def __call__(self, *args, **kwargs):
+				if len(args) + len(kwargs) != len(self._in_signature):
+					# TODO: Add type check for parameters
+					raise Exception("Wrong count of params for method '%s'." % method_name)
+				return method(*args, **kwargs)
+
+		self._unix_socket_methods[method_name] = wrapper(in_signature, out_signature)
+
+	def signal(self, method, out_signature):
+		if not ismethod(method):
+			raise Exception("Only bound methods can be exported.")
+		
+		method_name = method.__name__
+		if method_name in self._unix_socket_methods:
+			raise Exception("Method with this name (%s) is already exported." % method_name)
+		
+		class wrapper(object):
+			def __init__(self, out_signature):
+				self._out_signature = out_signature
+			
+			def __call__(self, *args, **kwargs):
+				return method(*args, **kwargs)
+		
+		self._unix_socket_methods[method_name] = wrapper(out_signature)
+		self._signals.add(method_name)
+
+	def send_signal(self, signal, *args, **kwargs):
+		if not signal in self._signals:
+			raise Exception("Signal '%s' doesn't exist." % signal)
+		for p in self._socket_signal_paths:
+			try:
+				s = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
```

Comment:
Why the change to STREAM? STREAM would have been perfect for a bidirectional channel, but you still use two separate unix sockets here.

Reply:
Oh, you did, just not for signals it seems.